### PR TITLE
Making `ConvBert Tokenizer` independent from `bert Tokenizer`

### PR DESCRIPTION
**Pull request by IMvision12**
# What does this PR do?

Fixes #19303

Added `BertTokenizer` class in tokenization_convbert.py and `BertTokenizerFast` in tokenization_convbert_fast.py

### CODE REVIEW

File: src/transformers/models/convbert/tokenization_convbert.py [hunk 1]
```diff
@@ -44,6 +49,467 @@
...
+                        f"Saving vocabulary to {vocab_file}: vocabulary indices are not consecutive."
+                        " Please check that the vocabulary is not corrupted!"
+                    )
+                    index = token_index
+                writer.write(token + "\n")
+                index += 1
+        return (vocab_file,)
+
+
+class BasicTokenizer(object):
```
Comment (sgugger) [inline src/transformers/models/convbert/tokenization_convbert.py:308]: Needs a copied from here!

File: src/transformers/models/convbert/tokenization_convbert.py [hunk 2]
```diff
@@ -44,6 +49,467 @@
...
+    """Runs basic whitespace cleaning and splitting on a piece of text."""
+    text = text.strip()
+    if not text:
+        return []
+    tokens = text.split()
+    return tokens
+
+
+# Copied from transformers.models.bert.tokenization_bert.BasicTokenizer
+class BertTokenizer(PreTrainedTokenizer):
```
Comment (sgugger) [inline src/transformers/models/convbert/tokenization_convbert.py:74]: This should become the `ConvBertTokenizer` class, not stay as a `BertTokenizer` :-)

File: src/transformers/models/convbert/tokenization_convbert_fast.py [hunk 3]
```diff
@@ -45,6 +50,154 @@
     "YituTech/conv-bert-small": {"do_lower_case": True},
 }
 
+"""Copied from transformers.models.bert.tokenization_bert.BertTokenizerFast"""
```
Comment (sgugger) [inline src/transformers/models/convbert/tokenization_convbert_fast.py:53]:
```suggestion
```

File: src/transformers/models/convbert/tokenization_convbert_fast.py [hunk 4]
```diff
@@ -45,6 +50,154 @@
...
+
+# Copied from transformers.models.bert.tokenization_bert_fast.BertTokenizerFast
+class BertTokenizerFast(PreTrainedTokenizerFast):
```
Comment (sgugger) [inline src/transformers/models/convbert/tokenization_convbert_fast.py:56]: Same comment here

**Review by sgugger — COMMENTED** [review]
Thanks for working on this! Left a couple of pointers on how to finish the work :-)

File: src/transformers/models/convbert/tokenization_convbert.py (hunk 2 above)
Comment (IMvision12) [reply src/transformers/models/convbert/tokenization_convbert.py:75]: should i delete `class ConvBertTokenizer(BertTokenizer):` at line 515?

**Comment by IMvision12** [comment]
Done! @sgugger

File: src/transformers/models/convbert/tokenization_convbert.py [hunk 5]
```diff
@@ -45,14 +50,465 @@
...
+def whitespace_tokenize(text):
+    """Runs basic whitespace cleaning and splitting on a piece of text."""
+    text = text.strip()
+    if not text:
+        return []
+    tokens = text.split()
+    return tokens
+
+
+class ConvBertTokenizer(PreTrainedTokenizer):
```
Comment (sgugger) [inline src/transformers/models/convbert/tokenization_convbert.py:75]: You will need the Copied from statement here at the top:
```suggestion
# Copied from transformers.models.bert.tokenization_bert.BertTokenizer with Bert->ConvBert,BERT->ConvBERT
class ConvBertTokenizer(PreTrainedTokenizer):
```

File: src/transformers/models/convbert/tokenization_convbert_fast.py [hunk 6]
```diff
@@ -46,17 +51,148 @@
 }
 
 
-class ConvBertTokenizerFast(BertTokenizerFast):
+# Copied from transformers.models.bert.tokenization_bert_fast.BertTokenizerFast
```
Comment (sgugger) [inline src/transformers/models/convbert/tokenization_convbert_fast.py:54]:
```suggestion
# Copied from transformers.models.bert.tokenization_bert_fast.BertTokenizerFast with Bert->ConvBERT, BERT->ConvBERT
```

**Review by sgugger — COMMENTED** [review]
Thanks for iterating! We're on the right path.
Now we just need to make sure the util scanning the copies is happy (e.g., the copies should match the original). I've left comments in that direction and you can check locally if the util is happy by running `make repo-consistency`

**Comment by IMvision12** [comment]
@sgugger Thanks for quick feeback,
For tokenization_convbert.py i have change the comment to `# Copied from transformers.models.bert.tokenization_bert.BertTokenizer with ConvBertTokenizer->BertTokenizer`

and for tokenization_convbert_fast.py i have change the comment to `# Copied from transformers.models.bert.tokenization_bert_fast.BertTokenizerFast with ConvBertTokenizerFast->ConvBertTokenizer
`

for convbert_fast `make repo-consistency` gives error : - src/transformers\models\convbert\tokenization_convbert_fast.py: copy does not match models.bert.tokenization_bert_fast.BertTokenizerFast at line 55

**Comment by sgugger** [comment]
You'll need to add broader patterns than just the full name of the tokenizer as BERT is used in the docstrings for instance. To see what the copy utils wants to modify, you can run `make fix-copies` locally :-)

**Comment by IMvision12** [comment]
After running `make fix-copies` it changes `slow_tokenizer_class = BertTokenizer` but it should be `ConvBertTokenizer`

**Comment by sgugger** [comment]
Yes, that's why I made the suggestions above.

**Comment by IMvision12** [comment]
So, do i need to copy `BertTokenizer` and `BertTokenizerFast` class to tokenization_convbert_fast.py as after adding those classes it passes all tests

**Comment by sgugger** [comment]
Just accept the suggestions above.

**Comment by IMvision12** [comment]
Done @sgugger

File: src/transformers/models/convbert/tokenization_convbert.py [hunk 7]
```diff
@@ -45,14 +50,466 @@
...
+    return tokens
+
+
+# Copied from transformers.models.bert.tokenization_bert.BertTokenizer with ConvBertTokenizer->BertTokenizer
+class ConvBertTokenizer(PreTrainedTokenizer):
     r"""
-    Construct a ConvBERT tokenizer. [`ConvBertTokenizer`] is identical to [`BertTokenizer`] and runs end-to-end
-    tokenization: punctuation splitting and wordpiece. Refer to superclass [`BertTokenizer`] for usage examples and
-    documentation concerning parameters.
+    Construct a BERT tokenizer. Based on WordPiece.
```
Comment (sgugger) [inline src/transformers/models/convbert/tokenization_convbert.py:78]: As said in the comment above, we need a BERT->ConvBERT replacement for the docstrings.

**Review by sgugger — COMMENTED** [review]
I'm not following what you're doing in the fast tokenizer file. The goal is to have ConvBertTokenizerFast stop depending on BertTokenizerFast by copying the code from that file. Not put the code of BertTokenizer there.

File: src/transformers/models/convbert/tokenization_convbert_fast.py [hunk 8]
```diff
@@ -46,17 +51,148 @@
 }
 
 
-class ConvBertTokenizerFast(BertTokenizerFast):
+# Copied from transformers.models.bert.tokenization_bert_fast.BertTokenizerFast with bert-base-cased->YituTech/conv-bert-base, Bert->ConvBert all-casing, ConvBertTokenizerFast->BertTokenizerFast
+class ConvBertTokenizerFast(PreTrainedTokenizerFast):
```
Comment (sgugger) [inline src/transformers/models/convbert/tokenization_convbert_fast.py:55]:
```suggestion
# Copied from transformers.models.bert.tokenization_bert_fast.BertTokenizerFast with bert-base-cased->YituTech/conv-bert-base, Bert->ConvBert, BERT->ConvBERT
class ConvBertTokenizerFast(PreTrainedTokenizerFast):
```
with all casing the BERT is converted to CONVBERT when its name is ConvBert. Also Bert->ConvBert will do the conversion of the class name :-)

File: src/transformers/models/convbert/tokenization_convbert.py [hunk 9]
```diff
@@ -45,14 +50,466 @@
...
+def whitespace_tokenize(text):
+    """Runs basic whitespace cleaning and splitting on a piece of text."""
+    text = text.strip()
+    if not text:
+        return []
+    tokens = text.split()
+    return tokens
+
+
+# Copied from transformers.models.bert.tokenization_bert.BertTokenizer with bert-base-cased->YituTech/conv-bert-base, ConvBertTokenizer->BertTokenizer, BERT->ConvBert
```
Comment (sgugger) [inline src/transformers/models/convbert/tokenization_convbert.py:75]:
```suggestion
# Copied from transformers.models.bert.tokenization_bert.BertTokenizer with bert-base-cased->YituTech/conv-bert-base, ConvBertTokenizer->BertTokenizer, BERT->ConvBERT
```
Nit: the name is ConvBERT

**Review by sgugger — APPROVED** [review]
Thanks a lot for iterating! This is almost ready to merge, just a couple of last nits.

**Review by sgugger — APPROVED** [review]
Thanks a lot for iterating on this PR! Looks perfect now.

**Pull request merged**